src/decryptionkey.rs

```diff
@@ -75,24 +105,61 @@ impl DecryptionKey {
                 lambda,
                 totient,
                 u,
+                n_inv,
+                p: p.clone(),
+                q: q.clone(),
             }),
         }
     }
 
     /// Reverse ciphertext to plaintext
     pub fn decrypt(&self, c: &Ciphertext) -> Option<Vec<u8>> {
-        if !mod_in(&c, &self.pk.nn) {
+        if !mod_in(c, &self.pk.nn) {
             return None;
         }
 
         // a = c^\lambda mod n^2
         let a = c.modpow(&self.lambda, &self.pk.nn);
+
         // ell = L(a, N)
-        self.pk.l(&a).map(|l| {
-            // m = lu = L(a)*u = L(c^\lamba*)u mod n
-            let m = l.modmul(&self.u, &self.pk.n);
-            m.to_bytes()
-        })
+        let ell = self.pk.l(&a)?;
+
+        // m = lu = L(a)*u = L(c^\lamba*)u mod n
+        let m = ell.modmul(&self.u, &self.pk.n);
+
+        Some(m.to_bytes())
+    }
+
+    /// Reverse ciphertext to plaintext
+    pub fn decrypt_unchecked(&self, c: &Ciphertext) -> BigNumber {
+        debug_assert!(mod_in(c, &self.pk.nn));
+
+        // a = c^\lambda mod n^2
+        let a = c.modpow(&self.lambda, &self.pk.nn);
+
+        // ell = L(a, N)
+        let ell = self.pk.l_unchecked(&a);
+
+        // m = lu = L(a)*u = L(c^\lamba*)u mod n
+        ell.modmul(&self.u, &self.pk.n)
+    }
+
+    /// Reverse ciphertext to plaintext and also retrieve the randomness
+    pub fn decrypt_with_randomness(&self, c: &Ciphertext) -> (BigNumber, BigNumber) {
+        let n = &self.pk.n;
+        let nn = &self.pk.nn;
+
+        let m = self.decrypt_unchecked(c);
+
+        // g^-m = (N + 1)^-m = 1 - m N (mod N^2)
+        let g_m_inv = BigNumber::one().modsub(&m.modmul(n, nn), nn);
+
+        // r^N = c . g^-m (mod N^2)
+        let r_n = c.modmul(&g_m_inv, nn);
+
+        let r = r_n.modpow(&self.n_inv, n);
+
+        (m, r)
     }
 
     /// Get this key's byte representation.
```

Comment:
Need to update this comment to reflect larger byte length.

Reply:
Good catch